src/services/completions.ts

```diff
@@ -1319,7 +1320,9 @@ function createCompletionEntry(
 
         awaitText += `(await ${propertyAccessToConvert.expression.getText()})`;
         insertText = needsConvertPropertyAccess ? `${awaitText}${insertText}` : `${awaitText}${insertQuestionDot ? "?." : "."}${insertText}`;
-        replacementSpan = createTextSpanFromBounds(propertyAccessToConvert.getStart(sourceFile), propertyAccessToConvert.end);
+        const isInAwaitExpression = tryCast(propertyAccessToConvert.parent, isAwaitExpression);

```

Comment:
`tryCast` is probably not the right thing to use here; I would just write `propertyAccessToConvert.parent.kind === SyntaxKind.AwaitExpression`.

However, I am a bit curious as to whether or not this code works if the parent is a parenthesized expression or something.

Reply:
> `tryCast` is probably not the right thing to use here; I would just write `propertyAccessToConvert.parent.kind === SyntaxKind.AwaitExpression`.

😑 Absolutely don't see any differences, but I'm okay with this.

> However, I am a bit curious as to whether or not this code works if the parent is a parenthesized expression or something.

Do you mean this (this is how it works)?

```ts
const a = (await fetch('.').blob)
const a = ((await fetch('.')).blob())
```

Reply:
Put another way, what if `propertyAccessToConvert.parent` is a `ParenthesizedExpression`? I believe we have helpers for skipping those. Or, maybe it doesn't matter and if it's parenthesized, the bug doesn't happen anyway. In either case, it's worth a test just to make sure.

Reply:
Yes, there is `walkUpParenthesizedExpressions`, but it gets a bit in following example:

```ts
await (x(1).foo)
```

I added support for this, but it is not shown in vscode (because of `range` of completion):

```ts
interface Foo { foo: string }
async function foo(x: (a: number) => Promise<Foo>)  {
    await (x(1).foo) // suggestion is here, but not shown
}
```

Also FYI this case is not handled for now (since parent is not await expression):

```ts
(await x(1).foo.test)
```

Reply:
Btw why user would ever write await (x(1).foo) ?

Reply:
If it's legal JS, someone will write it and then tell us we're broken. 😅

Reply:
That being said, I don't quite think what you have now is correct yet. If I have:

```
await x(1)./*2*/
```

Then the completion shouldn't go outside of the await, because it's on the call expression.

So I think maybe what you had before was fine (without the paren handling), and we just need explicit tests for each of these cases to make sure. That's the main thing I wanted considered.

Reply:
> That's the main thing I wanted considered

@jakebailey I'm so sorry, but I didn't get the point!

>  If I have:
> await x(1)./*2*/

So with this pr accepting `foo` completion from example above would result:

```ts
await x(1).foo /* -> */ (await x(1)).foo
```

Isn't that correct? I thought it was the whole point of linked issue. Can you please point me to exact problem and tests that I should add? Thx!

Reply:
Unless I'm totally missing something, I'm pretty sure that #48312 is saying that when you write `await x(1).` and accept a completion on property `foo`, the result should be `await x(1).foo`, which is to say, no parens.

I think I led you astray at the start as I was too focused on the `.parent` access, when the bug shouldn't appear if the parent is a paren already.

Reply:
@jakebailey

> Unless I'm totally missing something, I'm pretty sure that #48312 is saying that when you write `await x(1).` and accept a completion on property `foo`, the result should be `await x(1).foo`, which is to say, no parens.

But how it can make sense? Since then the foo completion would result a ts error `Property 'foo' does not exist on type 'Promise<Foo>'.`

What other people say (https://github.com/microsoft/TypeScript/issues/49423)

> It would be nice if VS Code knew to **reuse the existing await** in this context.

And see code from https://github.com/microsoft/TypeScript/issues/49423#issuecomment-1149154898

Reply:
FYI I reverted handling parenthesized expressions as VS Code doesn't give completions anyway.

> - Btw why user would ever write `await (x(1).foo)`?
> - If it's legal JS, someone will write it and then tell us we're broken. 😅

Instead of breaking this case we just don't handle it (keep behavior)

@jakebailey Can you say wether its good now? 😄

Reply:
I think it looks okay now, if the intent is to replace the completion fo `foo` on `await x(1).` with `(await x(1)).foo` exactly as the test shows. Sorry if I've really confused you; I don't think I understood the original bug report.

That being said, I'm not sure we have any precedence for this kind of completion either. Maybe @andrewbranch has a better opinion here.

One case I'm worried about is when the promise actually _does_ have properties. For example, the library `zx` returns promises with extra properties called [`ProcessPromise`](https://github.com/google/zx#processpromise). You can write something like `await $`echo hi`.quiet()`, and I'm thinking this PR would mean that when you're trying to write that code, you'll get some parens you shouldn't. So, I'm a little wary of adding this at all.

Reply:
This completion should only kick in when you try to access properties that can _only_ be accessed on the awaited type, so I don’t think it’s a problem here. It seems like your concern would apply to the original feature that inserts `await` too. I suspect it works correctly, but I’m not sure whether something extending `Promise` is tested.